LINUX/octeon_netmap.h

```diff
@@ -0,0 +1,498 @@
+
+		/* Enable POW interrupt when our port has at least one packet */
+		octeon_netmap_irq_enable(na, i, 1);
+
+		/* Mark the group as used */
+		set_bit(group, &netmap_receive_groups);
+	}
+
+	/* Configure port to use netmap receive group(s) */
+	pip_prt_tagx.u64 = cvmx_read_csr(CVMX_PIP_PRT_TAGX(port));
+	pip_prt_tagx.s.grptagbase = ona->base;
+	pip_prt_tagx.s.grptagmask = ~tag_mask;
+	pip_prt_tagx.s.grp = ona->base;
+	cvmx_write_csr(CVMX_PIP_PRT_TAGX(port), pip_prt_tagx.u64);
+
+	/* Set the default QOS level for the port */
+	pip_prt_cfgx.u64 = cvmx_read_csr(CVMX_PIP_PRT_CFGX(port));
+	pip_prt_cfgx.s.qos = 1;
+	cvmx_write_csr(CVMX_PIP_PRT_CFGX(port), pip_prt_cfgx.u64);
+
+	return 0;
+}
+
+static int octeon_netmap_disable_port(struct netmap_adapter *na)
+{
+	const int coreid = cvmx_get_core_num();
+	struct oct_nm_adapter *ona = (struct oct_nm_adapter *)na;
+	struct ifnet *ifp = na->ifp;
+	struct octeon_ethernet *priv = netdev_priv(ifp);
+	union cvmx_pip_prt_tagx pip_prt_tagx;
+	int port = priv->port;
+	int i;
+
+	nm_prdis("%s: Disable Port %d", ifp->name, port);
+
+	/* Configure port to use default receive group(s) */
+	pip_prt_tagx.u64 = cvmx_read_csr(CVMX_PIP_PRT_TAGX(port));
+	pip_prt_tagx.s.grptagbase = 0;
+	pip_prt_tagx.s.grptagmask = ~((1 << receive_group_order) - 1);
+	if (receive_group_order)
+		pip_prt_tagx.s.grp = 0;
+	else
+		pip_prt_tagx.s.grp = pow_receive_group;
+	cvmx_write_csr(CVMX_PIP_PRT_TAGX(port), pip_prt_tagx.u64);
+
+	/* Pass all queued packets to the default work queue */
+	for (i = 0; i < na->num_rx_rings; i++) {
+		while (1) {
+			u64 old_group_mask;
+			struct cvmx_wqe *work;
+			int group = ona->base + i;
+
+			old_group_mask =
+			    cvmx_read_csr(CVMX_POW_PP_GRP_MSKX(coreid));
+			cvmx_write_csr(CVMX_POW_PP_GRP_MSKX(coreid),
+				       (old_group_mask & ~0xFFFFull) | 1 <<
+				       group);
+			work = cvmx_pow_work_request_sync(CVMX_POW_NO_WAIT);
+			cvmx_write_csr(CVMX_POW_PP_GRP_MSKX(coreid),
+				       old_group_mask);
+			if (!work)
+				break;
+			octeon_netmap_submit_kernel(work);
+		}
+	}
+
+	/* Free the interrupt handlers */
+	for (i = 0; i < na->num_rx_rings; i++) {
+		int group = ona->base + i;
+
+		/* Disable interrupt and free the handler */
+		octeon_netmap_irq_enable(na, i, 0);
+		irq_set_affinity_hint(OCTEON_IRQ_WORKQ0 + group, NULL);
+		free_irq(OCTEON_IRQ_WORKQ0 + group, na);
+		clear_bit(group, &netmap_receive_groups);
+	}
+
+	return 0;
+}
+
+/*
+ * Register/unregister. We are already under netmap lock.
+ */
+static int octeon_netmap_reg(struct netmap_adapter *na, int onoff)
+{
+	int i, t;
+
+	nm_prdis("%s: %s", na->ifp->name, onoff ? "ON" : "OFF");
+
+	/* Enable or disable */
+	if (onoff) {
+		if (na->active_fds == 0) {
+			octeon_netmap_enable_port(na);
+		}
+		for_rx_tx(t) {
+			for (i = 0; i <= nma_get_nrings(na, t); i++) {
+				struct netmap_kring *kring = NMR(na, t)[i];
+				if (nm_kring_pending_on(kring)) {
+					kring->nr_mode = NKR_NETMAP_ON;
```

Comment:
For hardware NICs, this operation is usually performed by calling `netmap_reset` from the driver patch. You can look at intel drivers, for example, or vmxnet3.

Reply:
Done